drivers: sensor: add MAX31855

gwoplock opened the pull request:
add [MAX31855 ](https://www.analog.com/media/en/technical-documentation/data-sheets/max31855.pdf) cold-junction compensated thermocouple-to-digital converter sensor driver and sample

Reopening this PR under my account, as the previous PR got stale and was closed. (#22716)

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -68,6 +68,7 @@ add_subdirectory_ifdef(CONFIG_MAX17055		max17055)
 add_subdirectory_ifdef(CONFIG_MAX17262		max17262)
 add_subdirectory_ifdef(CONFIG_MAX30101		max30101)
 add_subdirectory_ifdef(CONFIG_MAX31875		max31875)
+add_subdirectory_ifdef(CONFIG_MAX31855		max31855)
```
Please move up to preserve alpha order

MaureenHelm commented on drivers/sensor/Kconfig:
```diff
@@ -177,6 +177,8 @@ source "drivers/sensor/max30101/Kconfig"
 
 source "drivers/sensor/max31875/Kconfig"
 
+source "drivers/sensor/max31855/Kconfig"
```
Please move up to preserve alpha order

MaureenHelm commented on drivers/sensor/max31855/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config MAX31855
+	bool "MAX31855 sensor"
+	depends on SPI
```
```suggestion
	default y
	depends on DT_HAS_MAXIM31855_ENABLED
	select SPI
```

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,163 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <drivers/sensor.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <sys/byteorder.h>
+#include <sys/__assert.h>
+#include <drivers/spi.h>
+#include <logging/log.h>
```
These all need `zephyr` in the path

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,163 @@
+	rx_buf.buf = buf;
+	rx_buf.len = size;
+
+	int ret = spi_read(data->spi, &data->spi_cfg, &rx);
```
Please use the _dt variants for SPI transactions

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,163 @@
+	struct max31855_data *data = dev->driver_data;
+	s32_t temp = sys_be32_to_cpu(data->data);
+
+	if (temp & (0x00010000)) {
```
Use the `BIT()` macro for the immediate

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,163 @@
+
+static inline int max31855_spi_init(struct max31855_data *data)
+{
+	data->spi = device_get_binding(DT_INST_0_MAXIM_MAX31855_BUS_NAME);
```
You can eliminate this runtime search by using `SPI_DT_SPEC_INST_GET` in the device config struct

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,163 @@
+	return 0;
+}
+
+static struct max31855_data max31855_data;
+
+DEVICE_AND_API_INIT(max31855, DT_INST_0_MAXIM_MAX31855_LABEL,
+		max31855_init, &max31855_data, NULL, POST_KERNEL,
+		CONFIG_SENSOR_INIT_PRIORITY, &max31855_api_funcs);
```
This needs to be updated to support multiple instances and to use `SENSOR_DEVICE_DT_INST_DEFINE`

MaureenHelm commented on drivers/sensor/max31855/max31855.h:
```diff
@@ -0,0 +1,21 @@
+#include <zephyr/types.h>
+#include <device.h>
+
+struct max31855_data {
+	struct device *spi;
+	struct spi_config spi_cfg;
+
+	/* Value */
+	s32_t data;
+};
```
It's not worth having a separate private include file for this, please move the struct definition to the .c file.

MaureenHelm commented on dts/bindings/sensor/maxim,max31855-spi.yaml:
```diff
@@ -0,0 +1,8 @@
+
+compatible: "maxim,max31855"
+
+include: spi-device.yaml
```
```suggestion
include: [sensor-device.yaml, spi-device.yaml]
```

MaureenHelm commented on samples/sensor/max31855/README.rst:
```diff
@@ -0,0 +1,35 @@
+Overview
+********
+
+This sample application demonstrates how to use the Maxim MAX31855
+cold-junction compensated thermocouple-to-digital converter to measure
+temperature from a thermocouple and the die temperature in °C.
```
We are moving away from driver-specific sensor sample applications. Please extend `samples/sensor/die_temp_polling` instead of adding a new sample

gwoplock replied:
If I'm not mistaken, I don't see any changes needed to the `die_temp_polling` sample to get it to support the sensor, and the other sensor channel is covered by the `thermometer` sample. Does removing the driver-specific sample suffice?

MaureenHelm replied:
Yes, thank you

MaureenHelm reviewed (changes requested):
Please add this sensor to `tests/drivers/build_all/sensor`

gwoplock committed:
drivers: sensor: add MAX31855

add MAX31855 cold-junction compensated thermocouple-to-digital
converter sensor driver and sample

Signed-off-by: Garrett Battaglia <garrett@garrettbattaglia.com>

greg-leach commented on drivers/sensor/max31855/Kconfig:
```diff
@@ -0,0 +1,12 @@
+# MAX31855 cold-junction compensated thermocpuple-to-digital converter configuration options
```
'thermocouple'

greg-leach commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,129 @@
+		}
+
+		val->val1 = temp / 4;
+		val->val2 = (temp % 4) * 250000;
```
Defines for magic numbers throughout?

greg-leach commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,129 @@
+	return 0;
+}
+
+#define MAX31855_INIT(n)							\
```
Spacing goes awry here.

greg-leach commented on dts/bindings/sensor/maxim,max31855-spi.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2020, Christian Hirsch
+# SPDX-License-Identifier: Apache-2.0
+
+description: MAX31855 SPI-based cold-junction compensated thermocpuple-to-digital converter.
```
'thermocouple'

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,132 @@
+
+	ret = spi_read_dt(&config->spi, &rx);
+	if (ret < 0) {
+		LOG_DBG("max31855_read FAIL %d", ret);
```
LOG_ERR

MaureenHelm commented on drivers/sensor/max31855/max31855.c:
```diff
@@ -0,0 +1,132 @@
+	.channel_get = max31855_channel_get,
+};
+
+int max31855_init(const struct device *dev)
```
Make this `static`

MaureenHelm reviewed (commented):
A couple of nits, but otherwise LGTM
